### ISSUE TITLE
Cannot compile with disabled precompiled headers

### ISSUE THREAD

**Issue by Bouncner** [hyrise/hyrise#2227]
When I disable precompiled headers (Ubuntu 19, GCC 9.2.1), I get the following error in the table key constraint:
```
[3/37] Building CXX object hyrise/src/lib/CMakeFiles/hyrise_impl.dir/storage/constraints/table_key_constraint.cpp.o
FAILED: hyrise/src/lib/CMakeFiles/hyrise_impl.dir/storage/constraints/table_key_constraint.cpp.o
ccache /usr/bin/g++  -DBOOST_THREAD_VERSION=5 -DHYRISE_DEBUG=1 -DHYRISE_NUMA_SUPPORT=1 -Dhyrise_impl_EXPORTS -I../hyrise/third_party/googletest/googletest/include -I../hyrise/third_party/sql-parser/src -I../hyrise/third_party/magic_enum/include -I../hyri
se/src/benchmarklib -I../hyrise/src/lib -I../hyrise/src/plugin -I. -I../hyrise/third_party/lz4/lib -I../hyrise/third_party/uninitialized_vector -isystem ../hyrise/third_party/benchmark/include -isystem ../hyrise/third_party/cpp-btree -isystem ../hyrise/t
hird_party/cxxopts/include -isystem ../hyrise/third_party/flat_hash_map -isystem ../hyrise/third_party/json -isystem ../hyrise/third_party/lz4 -isystem ../hyrise/third_party/zstd -isystem ../hyrise/third_party/tpcds-result-reproduction -isystem ../hyrise
/third_party/nlohmann_json/single_include -fdiagnostics-color=always -g -fopenmp-simd -O0 -fPIC   -DBOOST_ASIO_DISABLE_CONCEPTS -pthread -Wno-unknown-warning-option -Wall -Wextra -pedantic -Werror -Wno-unused-parameter -Wno-dollar-in-identifier-extension -Wno-unknown-pragmas -Wno-subobject-linkage -fPIC -std=gnu++2a -MD -MT hyrise/src/lib/CMakeFiles/hyrise_impl.dir/storage/constraints/table_key_constraint.cpp.o -MF hyrise/src/lib/CMakeFiles/hyrise_impl.dir/storage/constraints/table_key_constraint.cpp.o.d -o hyrise/src/lib/CMakeFiles/hyrise_impl.dir/storage/constraints/table_key_constraint.cpp.o -c ../hyrise/src/lib/storage/constraints/table_key_constraint.cpp
In file included from ../hyrise/src/lib/types.hpp:19,
                 from ../hyrise/src/lib/storage/constraints/abstract_table_constraint.hpp:5,
                 from ../hyrise/src/lib/storage/constraints/table_key_constraint.hpp:3,
                 from ../hyrise/src/lib/storage/constraints/table_key_constraint.cpp:1:
../hyrise/src/lib/storage/constraints/table_key_constraint.cpp: In member function ‘virtual bool opossum::TableKeyConstraint::_on_equals(const opossum::AbstractTableConstraint&) const’:
../hyrise/src/lib/utils/assert.hpp:63:30: error: the compiler can assume that the address of ‘table_constraint’ will never be NULL [-Werror=address]
   63 |   if (!static_cast<bool>(expr)) { \
      |                              ^
../hyrise/src/lib/utils/assert.hpp:75:32: note: in expansion of macro ‘Assert’
   75 | #define DebugAssert(expr, msg) Assert(expr, msg)
      |                                ^~~~~~
../hyrise/src/lib/storage/constraints/table_key_constraint.cpp:11:3: note: in expansion of macro ‘DebugAssert’
   11 |   DebugAssert(dynamic_cast<const AbstractTableConstraint*>(&table_constraint),
      |   ^~~~~~~~~~~
../hyrise/src/lib/utils/assert.hpp:63:3: error: nonnull argument ‘table_constraint’ compared to NULL [-Werror=nonnull-compare]
   63 |   if (!static_cast<bool>(expr)) { \
      |   ^~
../hyrise/src/lib/utils/assert.hpp:75:32: note: in expansion of macro ‘Assert’
   75 | #define DebugAssert(expr, msg) Assert(expr, msg)
      |                                ^~~~~~
../hyrise/src/lib/storage/constraints/table_key_constraint.cpp:11:3: note: in expansion of macro ‘DebugAssert’
   11 |   DebugAssert(dynamic_cast<const AbstractTableConstraint*>(&table_constraint),
      |   ^~~~~~~~~~~
At global scope:
cc1plus: error: unrecognized command line option ‘-Wno-dollar-in-identifier-extension’ [-Werror]
cc1plus: error: unrecognized command line option ‘-Wno-unknown-warning-option’ [-Werror]
```

**Comment by mrks** [comment]
1. Weird. I thought that PCH on GCC simply saved the internal state of the frontend. In that case, it shouldn't make a difference.
2. From a CI perspective, this is not surprising as we only have a clang non-PCH build. Feel free to add a gcc one.
3. The warning is correct - the dynamic_cast doesn't really convert anything.
4. I tried to build a SSCCE, but I can't even get `-Wnonnull-compare` to fire in a minimal setup.

**Comment by mrks** [comment]
This particular thing was fixed in #2248
